examples/gcf-pubsub-vm-delete-event-handler/main.py

```diff
@@ -0,0 +1,660 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License.  You may obtain a copy
+# of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+__author__ = ('Jeff McCune <jeff@openinfrastructure.co>, '
+              'Gary Larizza <gary@openinfrastructure.co>')
+
+import base64
+import json
+import logging
+import os
+import google.cloud.logging
+from enum import Enum
+from google.cloud.logging.resource import Resource
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from typing import List
+
+
+class RuntimeState:
+    """Stores App instance for the lifetime of the process"""
+    pass
+
+
+RuntimeState.app = None
+
+
+class Result(Enum):
+    """The overall result of the DNS cleanup for user reporting
+
+    An OK result indicates the cleanup completed normally.
+
+    A NOT_PROCESSED results is likely a result of losing the race against the
+    VM delete operation and is intended to signal to the user they may need to
+    cleanup DNS records using another mechanism (e.g. manually).
+    """
+    OK = 0
+    NOT_PROCESSED = 1
+
+
+class Severity(Enum):
+    """Stackdriver severity levels
+
+    See https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+    """
+    DEFAULT = 0
+    DEBUG = 100
+    INFO = 200
+    NOTICE = 300
+    WARNING = 400
+    ERROR = 500
+    CRITICAL = 600
+    ALERT = 700
+    EMERGENCY = 800
+
+
+class Detail(Enum):
+    """Detailed results of the cleanup
+
+    A LOST_RACE result indicates user intervention is necessary.
+    """
+    NO_OP = 0
+    NO_MATCHES = 1
+    RR_DELETED = 2
+    VM_NO_IP = 3
+    IGNORED_EVENT = 4
+    LOST_RACE = 5
+    RR_MISMATCH = 6
+    RR_NOT_A_RECORD = 7
+    RR_NAME_MISMATCH = 8
+    RR_IP_MISMATCH = 9
+
+
+class StructuredLog(object):
+    """Base class to report structured log events
+
+    Sub classes are expected to override MESSAGE (str), RESULT (Result), and
+    DETAIL (Detail).
+
+    """
+    MESSAGE = '{} mark'
```

Comment:
what is the  `mark` for?

Reply:
It comes from the convention of logging nothing having the behavior of a "mark" syslog entry marking the fact that nothing was logged.

Sub classes should always override this attribute, so I'll remove it entirely.  With the other changes, there will be an error when `self.MESSAGE.format` is called in the sub-class if it's not overridden.

Reply:
I think "mark" in stackdriver is just log noise. 

You can specify an abstract attribute like so 
```python
   # In base class
    @property
    @abstractmethod
    def MESSAGE(self):
         raise NotImplementedError
...
    # In subclass
    MESSAGE = 'string for formatting {}'
```

Reply:
Fixed in ecde344...5eef9da